src/main/java/org/wildfly/security/sasl/oauth2/OAuth2SaslClientFactory.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.sasl.oauth2;
+
+import org.kohsuke.MetaInfServices;
+import org.wildfly.security.mechanism.oauth2.OAuth2Client;
+import org.wildfly.security.sasl.util.SaslMechanismInformation;
+
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslClientFactory;
+import javax.security.sasl.SaslException;
+import java.util.Map;
+
+/**
+ * A {@link SaslClientFactory} for OAuth2.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+@MetaInfServices(value = SaslClientFactory.class)
+public final class OAuth2SaslClientFactory implements SaslClientFactory {
+
+    public SaslClient createSaslClient(final String[] mechanisms, final String authorizationId, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        for (String mechanism : mechanisms) {
+            if (SaslMechanismInformation.Names.OAUTHBEARER.equals(mechanism)) {
+                return new OAuth2SaslClient(mechanism, protocol, serverName, cbh, authorizationId, new OAuth2Client(mechanism, authorizationId, cbh));
+            }
+        }
+
+        return null;
+    }
+
+    public String[] getMechanismNames(final Map<String, ?> props) {
```

Comment:
I see your point. I don't think that will actually mitigate the risks of a replay attack, given that servers are usually protected against this type of attack. At least they should be and consider http://www.rfc-base.org/txt/rfc-6819.txt.

But I agree with you that we can give users an option to disable this client if they want to and avoid what you just described.

Which property would be appropriate for this case ? Should I create a new one in ```WildFlySasl``` ?